src/reyna/MessageProvider/BatchProvider.cs

```diff
@@ -0,0 +1,116 @@
+﻿namespace Reyna
+{
+    using System;
+    using System.Net;
+    using Reyna.Interfaces;
+
+    internal class BatchProvider : IMessageProvider
+    {
+        private const string PeriodicBackoutCheckTAG = "BatchProvider";
+
+        public BatchProvider(IRepository repository, IPeriodicBackoutCheck periodicBackoutCheck)
+        {
+            this.Repository = repository;
+            this.BatchConfiguration = new BatchConfiguration();
+            this.PeriodicBackoutCheck = periodicBackoutCheck;
+        }
+
+        public bool CanSend
+        {
+            get
+            {
+                long interval = (long)(this.BatchConfiguration.SubmitInterval * 0.9);
+                if (this.PeriodicBackoutCheck.IsTimeElapsed(PeriodicBackoutCheckTAG, interval))
+                {
+                    return true;
+                }
+
+                return this.Repository.AvailableMessagesCount >= this.BatchConfiguration.BatchMessageCount;
+            }
+        }
+
+        internal IBatchConfiguration BatchConfiguration { get; set; }
+
+        internal IPeriodicBackoutCheck PeriodicBackoutCheck { get; set; }
+
+        private IRepository Repository { get; set; }
+
+        private bool BatchDeleted { get; set; }
+
+        public IMessage GetNext()
+        {
+            var message = this.Repository.Get();
+            var batch = new Batch();
+
+            WebHeaderCollection headers = null;
+            var count = 0;
+            long size = 0;
+            long maxMessagesCount = this.BatchConfiguration.BatchMessageCount;
+            long maxBatchSize = this.BatchConfiguration.BatchMessagesSize;
+            while (message != null && count < maxMessagesCount && size < maxBatchSize)
+            {
+                headers = message.Headers;
+                batch.Add(message);
+                
+                size = this.GetSize(batch);
+                count++;
+                message = this.Repository.GetNextMessageAfter(message.Id);
+            }
```

Comment:
same logic in android.

we check during adding the message to the batch.

if batch size exceeds we remove the last message